src/fr-webauthn/index.ts

```diff
@@ -0,0 +1,309 @@
+   *
+   * @param credential The credential to convert
+   * @return The outcome string
+   */
+  public static getAuthenticationOutcome(credential: PublicKeyCredential | null): string {
+    if (!window.PublicKeyCredential) {
+      return WebAuthnOutcome.Unsupported;
+    }
+
+    try {
+      if (credential === null) {
+        throw new Error('No credential provided');
+      }
+
+      const clientDataJSON = getClientDataJson(credential);
+      const assertionResponse = credential.response as AuthenticatorAssertionResponse;
+      const authenticatorData = new Int8Array(assertionResponse.authenticatorData).toString();
+      const signature = new Int8Array(assertionResponse.signature).toString();
+
+      return `${clientDataJSON}::${authenticatorData}::${signature}::${credential.id}`;
+    } catch (error) {
+      return this.getErrorOutcome(error);
+    }
+  }
+
+  /**
+   * Retrieves the credential from the browser Web Authentication API.
+   *
+   * @param options The public key options associated with the request
+   * @return The credential
+   */
+  public static async getRegistrationCredential(
+    options: PublicKeyCredentialCreationOptions,
+  ): Promise<PublicKeyCredential | null> {
+    const credential = await navigator.credentials.create({ publicKey: options });
+    return credential as PublicKeyCredential;
+  }
+
+  /**
+   * Converts a registration credential into the outcome expected by OpenAM.
+   *
+   * @param credential The credential to convert
+   * @return The outcome string
+   */
+  public static getRegistrationOutcome(credential: PublicKeyCredential | null): string {
+    if (!window.PublicKeyCredential) {
+      return WebAuthnOutcome.Unsupported;
+    }
+
+    try {
+      if (credential === null) {
+        throw new Error('No credential provided');
+      }
+
+      const clientDataJSON = getClientDataJson(credential);
+      const attestationResponse = credential.response as AuthenticatorAttestationResponse;
+      const attestationObject = new Int8Array(attestationResponse.attestationObject).toString();
+      return `${clientDataJSON}::${attestationObject}::${credential.id}`;
+    } catch (error) {
+      return this.getErrorOutcome(error);
+    }
+  }
+
+  /**
+   * Converts authentication tree metadata into options required by the browser
+   * Web Authentication API.
+   *
+   * @param metadata The metadata provided in the authentication tree MetadataCallback
+   * @return The Web Authentication API request options
+   */
+  public static createAuthenticationPublicKey(
+    metadata: WebAuthnAuthenticationMetadata,
+  ): PublicKeyCredentialRequestOptions {
+    const { acceptableCredentials, challenge, relyingPartyId, timeout } = metadata;
+    const rpId = parseRelyingPartyId(relyingPartyId);
+
+    return {
+      allowCredentials: parseCredentials(acceptableCredentials),
+      challenge: Uint8Array.from(atob(challenge), (c) => c.charCodeAt(0)).buffer,
+      timeout,
+      ...(rpId && { rpId }),
```

Comment:
This is quite clever. Though, I'm not a huge fan of using clever code as it sets a dangerous precedent. I'm guessing you wanted a super succinct way to only include the property of `rpId` if it exists with as few variable declarations as possible?

Reply:
It's more about reducing overly-explicit conditionals so the function is easier to grok.  I wouldn't refer to this as "clever" because this style is used all the time in React JSX, and it's pretty easy to understand.  Another thing for the style guide.